src/annotator/anchoring/pdf.js

```diff
@@ -0,0 +1,430 @@
+  // offset value.
+  //
+  // When we find the page the offset is on, we return an object containing the
+  // page index, the offset at the start of that page, and the textContent of
+  // that page.
+  //
+  // To understand this a little better, here's a worked example. Imagine a
+  // document with the following page lengths:
+  //
+  //    Page 0 has length 100
+  //    Page 1 has length 50
+  //    Page 2 has length 50
+  //
+  // Then here are the pages that various offsets are found on:
+  //
+  //    offset | index
+  //    --------------
+  //    0      | 0
+  //    99     | 0
+  //    100    | 1
+  //    101    | 1
+  //    149    | 1
+  //    150    | 2
+  const count = textContent => {
+    const lastPageIndex = PDFViewerApplication.pdfViewer.pagesCount - 1;
+    if (total + textContent.length > offset || index === lastPageIndex) {
+      offset = total;
+      return Promise.resolve({ index, offset, textContent });
+    } else {
+      ++index;
+      total += textContent.length;
+      return getPageTextContent(index).then(count);
+    }
+  };
+
+  return getPageTextContent(0).then(count);
+}
+
+/**
+ * Map a position anchor to a region within the text layer of a rendered page.
+ *
+ * If the page is not rendered, a placeholder element is created and the
+ * returned range is the placeholder.
+ *
+ * @param {PDFPageView} page - The PDF.js viewer page
+ * @param {TextPositionAnchor} anchor - Anchor to locate in page
+ * @param {Object} options - Options for `anchor.toSelector`
+ * @return {Range}
+ */
+function anchorByPosition(page, anchor, options) {
+  let renderingDone = false;
+  if (page.textLayer) {
+    renderingDone = page.textLayer.renderingDone;
+  }
+  if (page.renderingState === RenderingStates.FINISHED && renderingDone) {
+    // The page has been rendered. Use HTML anchoring to locate the quote in
+    // the text layer.
+    const root = page.textLayer.textLayerDiv;
+    const selector = anchor.toSelector(options);
+    return html.anchor(root, [selector]);
+  }
+
+  // The page has not been rendered yet. Create a placeholder element and
+  // anchor to that instead.
+  const div = page.div || page.el;
+  let placeholder = div.getElementsByClassName('annotator-placeholder')[0];
+  if (!placeholder) {
+    placeholder = document.createElement('span');
+    placeholder.classList.add('annotator-placeholder');
+    placeholder.textContent = 'Loading annotations…';
+    div.appendChild(placeholder);
+  }
+  const range = document.createRange();
+  range.setStartBefore(placeholder);
+  range.setEndAfter(placeholder);
+  return range;
+}
+
+/**
+ * Search for a quote in the given pages.
+ *
+ * @param {number[]} pageIndexes - Pages to search in priority order
+ * @param {TextQuoteSelector} quoteSelector
+ * @return {Promise<Range>} Location of quote
+ */
+function findInPages(pageIndexes, quoteSelector, positionHint) {
+  if (pageIndexes.length === 0) {
+    return Promise.reject(new Error('Quote not found'));
+  }
+
```

Comment:
Internal comments for this function would be super helpful in the future. I'm not going to deeply analyze what it's doing for now...